src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,583 @@
+ * LQP that use correlated parameters.
+ */
+std::pair<bool, size_t> assess_correlated_parameter_usage(
+    const std::shared_ptr<AbstractLQPNode>& lqp,
+    const std::map<ParameterID, std::shared_ptr<AbstractExpression>>& parameter_mapping) {
+  bool optimizable = true;
+  size_t correlated_predicate_node_count = 0;
+  visit_lqp(lqp, [&](const auto& node) {
+    if (!optimizable) {
+      return LQPVisitation::DoNotVisitInputs;
+    }
+
+    if (uses_correlated_parameters(node, parameter_mapping)) {
+      if (node->type == LQPNodeType::Predicate) {
+        ++correlated_predicate_node_count;
+      } else {
+        optimizable = false;
+        return LQPVisitation::DoNotVisitInputs;
+      }
+    }
+
+    return LQPVisitation::VisitInputs;
+  });
+
+  return {!optimizable, correlated_predicate_node_count};
+}
+
+/**
+ * Tries to safely extract a join predicate from a correlated predicate node.
+ *
+ * Returns a binary predicate expression where the left operand is always the expression associated with the correlated
+ * parameter (and thus a column from the left subtree) and the right operand a column from the subqueries LQP.
+ */
+std::shared_ptr<BinaryPredicateExpression> try_to_extract_join_predicate(
+    const std::shared_ptr<PredicateNode>& predicate_node,
+    const std::map<ParameterID, std::shared_ptr<AbstractExpression>>& parameter_mapping, bool is_below_aggregate) {
+  // Check for the type of expression first. Note that we are not concerned with predicates of other forms using
+  // correlated parameters here. We check for parameter usages that prevent optimization later in
+  // contains_unoptimizable_correlated_parameter_usages.
+  if (predicate_node->predicate()->type != ExpressionType::Predicate) {
+    return nullptr;
+  }
+
+  const auto& predicate_expression = std::static_pointer_cast<AbstractPredicateExpression>(predicate_node->predicate());
+
+  // Joins only support these six binary predicates. We rely on PredicateSplitUpRule having split up ANDed chains of
+  // such predicates previously, so that we can process them separately.
+  auto predicate_condition = predicate_expression->predicate_condition;
+  switch (predicate_condition) {
+    case PredicateCondition::Equals:
+    case PredicateCondition::NotEquals:
+    case PredicateCondition::LessThan:
+    case PredicateCondition::LessThanEquals:
+    case PredicateCondition::GreaterThan:
+    case PredicateCondition::GreaterThanEquals:
+      break;
+    default:
+      return nullptr;
+  }
+
+  // We can currently only pull equals predicates above aggregate nodes (by grouping by the column that the predicate
+  // compares with). The other predicate types could be supported but would require more sophisticated reformulations.
+  if (is_below_aggregate && predicate_condition != PredicateCondition::Equals) {
+    return nullptr;
+  }
+
+  // Check that one side of the expression is a correlated parameter and the other a column expression of the LQP below
```

Comment:
If I understand the code correctly we just make sure we get one correlated parameter and then assume the other side must be "a column expression of the LQP below the predicate node". This is however not checked, contrary to what this comment implies, right? Also, what if left_side and right_side are both correlated parameters?